src/charm.py

```diff
@@ -31,12 +31,62 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.get_primary_action,
+                               self._on_get_primary_action)
+
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
```

Comment:
In this relation joined event a new Mongodb peer unit has joined so why not just add it to the replica set, without bothering to check if all planned units have joined ? In the future when additional peer units join the relation the charm will again get relation joined events for them too.